File: backend/schemas/jobs.py
from __future__ import annotations

from pydantic import BaseModel, Field

from backend.domain.entities import Job, JobEvent


class RelatedJobResponse(BaseModel):
    id: str
    status: str
    action: str
    created_at: str | None


class JobResponse(BaseModel):
    id: str
    type: str
    status: str
    artist_id: str | None
    input_user_id: str | None
    input_artwork_id: str | None
    options: dict[str, object]
    workflow_run_id: str | None
    workflow_item_id: int | None
    workflow_source: str | None
    total_files: int
    completed_files: int
    skipped_files: int
    failed_files: int
    cancel_requested: bool
    created_at: str | None
    started_at: str | None
    finished_at: str | None
    error_message: str | None
    related_jobs: list[RelatedJobResponse] = Field(default_factory=list)


class JobListResponse(BaseModel):
    items: list[JobResponse]
    total: int


class JobEventResponse(BaseModel):
    id: int | None
    job_id: str
    level: str
    message: str
    payload: dict[str, object] | None
    created_at: str | None


class JobEventListResponse(BaseModel):
    items: list[JobEventResponse]
    total: int


class JobDetailResponse(JobResponse):
    events: list[JobEventResponse]


class JobCancelResponse(BaseModel):
    job_id: str
    status: str
    cancel_requested: bool


class JobActionResponse(BaseModel):
    job_id: str
    status: str
    source_job_id: str
    action: str


class JobBulkCancelRequest(BaseModel):
    job_ids: list[str]


class JobBulkActionError(BaseModel):
    job_id: str
    message: str


class JobBulkCancelResponse(BaseModel):
    cancelled: list[JobCancelResponse]
    errors: list[JobBulkActionError]


class JobQueueStateResponse(BaseModel):
    paused: bool


class JobStreamMessage(BaseModel):
    type: str
    job_id: str
    status: str
    total_files: int
    completed_files: int
    skipped_files: int
    failed_files: int
    message: str
    created_at: str | None


def job_response(job: Job, *, related_jobs: list[Job] | None = None) -> JobResponse:
    return JobResponse(
        id=job.id,
        type=job.type,
        status=job.status,
        artist_id=job.artist_id,
        input_user_id=job.input_user_id,
        input_artwork_id=job.input_artwork_id,
        options=job.options,
        workflow_run_id=job.workflow_run_id,
        workflow_item_id=job.workflow_item_id,
        workflow_source=job.workflow_source,
        total_files=job.total_files,
        completed_files=job.completed_files,
        skipped_files=job.skipped_files,
        failed_files=job.failed_files,
        cancel_requested=job.cancel_requested,
        created_at=job.created_at,
        started_at=job.started_at,
        finished_at=job.finished_at,
        error_message=job.error_message,
        related_jobs=[related_job_response(child) for child in related_jobs or []],
    )


def related_job_response(job: Job) -> RelatedJobResponse:
    action = job.options.get("job_action", "retry")
    return RelatedJobResponse(
        id=job.id,
        status=job.status,
        action=str(action),
        created_at=job.created_at,
    )


def job_event_response(event: JobEvent) -> JobEventResponse:
    return JobEventResponse(
        id=event.id,
        job_id=event.job_id,
        level=event.level,
        message=event.message,
        payload=event.payload,
        created_at=event.created_at,
    )

File: frontend/src/api/jobs.ts
import { apiRequest } from "./client";
import type { FailureDetail } from "./scheduledTasks";

export type JobStatus = "inactive" | "queued" | "running" | "completed" | "failed" | "cancelled";

export type RelatedJob = {
  id: string;
  status: JobStatus;
  action: string;
  created_at: string | null;
};

export type Job = {
  id: string;
  type: string;
  status: JobStatus;
  artist_id: string | null;
  input_user_id: string | null;
  input_artwork_id: string | null;
  options: Record<string, unknown>;
  workflow_run_id: string | null;
  workflow_item_id: number | null;
  workflow_node_run_id: number | null;
  workflow_source: string | null;
  total_files: number;
  completed_files: number;
  skipped_files: number;
  failed_files: number;
  cancel_requested: boolean;
  created_at: string | null;
  started_at: string | null;
  finished_at: string | null;
  error_message: string | null;
  failure: FailureDetail | null;
  related_jobs: RelatedJob[];
};

export type JobEvent = {
  id: number | null;
  job_id: string;
  level: "debug" | "info" | "warning" | "error" | string;
  message: string;
  payload: Record<string, unknown> | null;
  created_at: string | null;
};

export type JobDetail = Job & {
  events: JobEvent[];
};

export type JobListResponse = {
  items: Job[];
  total: number;
};

export type JobCancelResponse = {
  job_id: string;
  status: JobStatus;
  cancel_requested: boolean;
};

export type JobActionResponse = {
  job_id: string;
  status: JobStatus;
  source_job_id: string;
  action: string;
};

export type JobBulkCancelResponse = {
  cancelled: JobCancelResponse[];
  errors: Array<{ job_id: string; message: string }>;
};

export type JobQueueState = {
  paused: boolean;
};

export type JobStreamMessage = {
  type: "job_progress" | "job_event" | "job_completed" | "job_failed" | "job_cancelled" | string;
  job_id: string;
  status: JobStatus;
  total_files: number;
  completed_files: number;
  skipped_files: number;
  failed_files: number;
  message: string;
  created_at: string | null;
};

export type ListJobsParams = {
  status?: JobStatus | "";
  limit?: number;
  offset?: number;
};

export function listJobs(params: ListJobsParams = {}): Promise<JobListResponse> {
  const search = new URLSearchParams();
  if (params.status) {
    search.set("status", params.status);
  }
  if (params.limit) {
    search.set("limit", String(params.limit));
  }
  if (params.offset) {
    search.set("offset", String(params.offset));
  }
  const query = search.toString();
  return apiRequest<JobListResponse>(`/jobs${query ? `?${query}` : ""}`);
}

export function getJob(jobId: string): Promise<JobDetail> {
  return apiRequest<JobDetail>(`/jobs/${jobId}`);
}

export function cancelJob(jobId: string): Promise<JobCancelResponse> {
  return apiRequest<JobCancelResponse>(`/jobs/${jobId}/cancel`, {
    method: "POST"
  });
}

export function retryJob(jobId: string): Promise<JobActionResponse> {
  return apiRequest<JobActionResponse>(`/jobs/${jobId}/retry`, {
    method: "POST"
  });
}

export function rerunJob(jobId: string): Promise<JobActionResponse> {
  return apiRequest<JobActionResponse>(`/jobs/${jobId}/rerun`, {
    method: "POST"
  });
}

export function bulkCancelJobs(jobIds: string[]): Promise<JobBulkCancelResponse> {
  return apiRequest<JobBulkCancelResponse>("/jobs/bulk-cancel", {
    method: "POST",
    body: { job_ids: jobIds }
  });
}

export function getJobQueueState(): Promise<JobQueueState> {
  return apiRequest<JobQueueState>("/jobs/queue");
}

export function pauseJobQueue(): Promise<JobQueueState> {
  return apiRequest<JobQueueState>("/jobs/queue/pause", {
    method: "POST"
  });
}

export function resumeJobQueue(): Promise<JobQueueState> {
  return apiRequest<JobQueueState>("/jobs/queue/resume", {
    method: "POST"
  });
}

export function listJobEvents(jobId: string): Promise<JobEvent[]> {
  return apiRequest<JobEvent[]>(`/jobs/${jobId}/events`);
}
